File: SingletonPrototype/Java/src/Ex1.java
import java.util.*;

// Paleta de cores
class Cores {
    public static final String RESET = "\u001B[0m";
    public static final String PINK = "\u001B[35m";
    public static final String GREEN = "\u001B[32m";
    public static final String PURPLE = "\u001B[38;5;93m";
    public static final String BOLD = "\u001B[1m";
    public static final String WHITE = "\u001B[97m";
}

// Classe Aluno (PROTOTYPE)
class Aluno implements Cloneable {
    int id;
    String nome;
    int idade;
    String curso;
    String unidade;
    String periodo;

    public Aluno(int id, String nome, int idade, String curso, String unidade, String periodo) {
        this.id = id;
        this.nome = nome;
        this.idade = idade;
        this.curso = curso;
        this.unidade = unidade;
        this.periodo = periodo;
    }

    public Aluno clone() {
        return new Aluno(id, nome, idade, curso, unidade, periodo);
    }
}

// Classe GerenciadorAlunos (SINGLETON)
class GerenciadorAlunos {
    private static GerenciadorAlunos instance;
    private Map<Integer, Aluno> alunos;

    private GerenciadorAlunos() {
        alunos = new HashMap<>();
    }

    public static GerenciadorAlunos getInstance() {
        if (instance == null) {
            instance = new GerenciadorAlunos();
        }
        return instance;
    }

    public void addAluno(Aluno aluno) {
        alunos.put(aluno.id, aluno);
        System.out.println(Cores.GREEN + "Aluno " + aluno.nome + " adicionado com sucesso!" + Cores.RESET);
    }

    public void listarAlunos() {
        System.out.println(Cores.PINK + Cores.BOLD + "\nLISTA DE ALUNOS" + Cores.RESET);

        System.out.println(Cores.PINK + "+-----+------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);

        System.out.printf(Cores.PINK + Cores.BOLD + "| %-3s | %-10s | %-5s | %-30s | %-15s | %-10s |\n" + Cores.RESET,
                "ID", "NOME", "IDADE", "CURSO", "UNIDADE", "PERIODO");

        System.out.println(Cores.PINK + "+-----+------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);

        for (Aluno a : alunos.values()) {
            System.out.printf(Cores.WHITE + "| %-3d | %-10s | %-5d | %-30s | %-15s | %-10s |\n" + Cores.RESET,
                    a.id, a.nome, a.idade, a.curso, a.unidade, a.periodo);
        }

        System.out.println(Cores.PINK + "+-----+------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);
    }
}

// Classe principal
public class Ex1 {
    public static void main(String[] args) {

        GerenciadorAlunos g1 = GerenciadorAlunos.getInstance();
        GerenciadorAlunos g2 = GerenciadorAlunos.getInstance();

        System.out.println(Cores.PINK + Cores.BOLD + "TESTE SINGLETON" + Cores.RESET);

        // Aluno Protótipo
        Aluno alunoPrototipo = new Aluno(
                0,
                "ALUNO PADRAO",
                0,
                "Desenvolvimento de Software",
                "Fatec Diadema",
                "Matutino"
        );

        // Clones
        Aluno aluno1 = alunoPrototipo.clone();
        aluno1.id = 1;
        aluno1.nome = "Joao";
        aluno1.idade = 20;

        Aluno aluno2 = alunoPrototipo.clone();
        aluno2.id = 2;
        aluno2.nome = "Maria";
        aluno2.idade = 22;

        g1.addAluno(aluno1);
        g2.addAluno(aluno2);

        g1.listarAlunos();

        // Protótipo original
        System.out.println(Cores.PURPLE + Cores.BOLD + "\nALUNO PROTOTIPO ORIGINAL" + Cores.RESET);

        System.out.println(Cores.PURPLE + "+-----+-----------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);

        System.out.printf(Cores.PURPLE + "| %-3s | %-15s | %-5s | %-30s | %-15s | %-10s |\n" + Cores.RESET,
                "ID", "NOME", "IDADE", "CURSO", "UNIDADE", "PERIODO");

        System.out.println(Cores.PURPLE + "+-----+-----------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);

        System.out.printf(Cores.WHITE + "| %-3d | %-15s | %-5d | %-30s | %-15s | %-10s |\n" + Cores.RESET,
                alunoPrototipo.id,
                alunoPrototipo.nome,
                alunoPrototipo.idade,
                alunoPrototipo.curso,
                alunoPrototipo.unidade,
                alunoPrototipo.periodo);

        System.out.println(Cores.PURPLE + "+-----+-----------------+-------+--------------------------------+-----------------+------------+" + Cores.RESET);
    }
}

File: SingletonPrototype/Python/Ex1.py
# Paleta de cores
class Cores:
    reset = "\x1b[0m"
    pink = "\x1b[35m"   
    green = "\x1b[32m"  
    purple = "\x1b[38;5;93m"
    black = "\x1b[1m" 
    bold = "\033[1m"

# Classe Aluno (PROTOTYPE)
class Aluno:
    def __init__(self, id, nome, idade, curso, unidade, periodo):
        self.id = id
        self.nome = nome
        self.idade = idade
        self.curso = curso
        self.unidade = unidade
        self.periodo = periodo

    def clone(self):
        return Aluno(
            self.id,
            self.nome,
            self.idade,
            self.curso,
            self.unidade,
            self.periodo
        )


# Classe GerenciadorAlunos (SINGLETON)
class GerenciadorAlunos:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(GerenciadorAlunos, cls).__new__(cls)
            cls._instance.alunos = {}
        return cls._instance

    def add_aluno(self, aluno):
        self.alunos[aluno.id] = aluno
        print(f"{Cores.green}✔ Aluno {aluno.nome} adicionado com sucesso!{Cores.reset}")

    def listar_alunos(self):
        largura = 105

        print(f"{Cores.pink}{Cores.bold}\n📚 LISTA DE ALUNOS{Cores.reset}")

        print(f"{Cores.pink}+{'='*largura}+{Cores.reset}")

        print(f"{Cores.pink}{Cores.bold}| "
              f"{'ID':<5} | {'NOME':<15} | {'IDADE':<7} | {'CURSO':<30} | {'UNIDADE':<20} | {'PERÍODO':<10}"
              f" |{Cores.reset}")

        print(f"{Cores.pink}+{'-'*largura}+{Cores.reset}")

        for aluno in self.alunos.values():
            print(f"{Cores.black}| "
                  f"{aluno.id:<5} | {aluno.nome:<15} | {aluno.idade:<7} | "
                  f"{aluno.curso:<30} | {aluno.unidade:<20} | {aluno.periodo:<10}"
                  f" |{Cores.reset}")

        print(f"{Cores.pink}+{'='*largura}+{Cores.reset}")


# Cliente
g1 = GerenciadorAlunos()
g2 = GerenciadorAlunos()

print(f"{Cores.pink}{Cores.bold}🔍 TESTE SINGLETON{Cores.reset}")

# Aluno Protótipo
aluno_prototipo = Aluno(
    0,
    "ALUNO PADRAO",
    0,
    "Desenvolvimento de Software",
    "Fatec Diadema",
    "Matutino"
)

# Clones
aluno1 = aluno_prototipo.clone()
aluno1.id = 1
aluno1.nome = "João"
aluno1.idade = 20

aluno2 = aluno_prototipo.clone()
aluno2.id = 2
aluno2.nome = "Maria"
aluno2.idade = 22

g1.add_aluno(aluno1)
g2.add_aluno(aluno2)

g1.listar_alunos()


largura = 105

print(f"{Cores.purple}{Cores.bold}\n📌 ALUNO PROTÓTIPO ORIGINAL{Cores.reset}")

print(f"{Cores.purple}+{'='*largura}+{Cores.reset}")

print(f"{Cores.purple}| "
      f"{'ID':<5} | {'NOME':<15} | {'IDADE':<7} | {'CURSO':<30} | {'UNIDADE':<20} | {'PERÍODO':<10}"
      f" |{Cores.reset}")

print(f"{Cores.purple}+{'-'*largura}+{Cores.reset}")

print(f"{Cores.black}| "
      f"{aluno_prototipo.id:<5} | {aluno_prototipo.nome:<15} | {aluno_prototipo.idade:<7} | "
      f"{aluno_prototipo.curso:<30} | {aluno_prototipo.unidade:<20} | {aluno_prototipo.periodo:<10}"
      f" |{Cores.reset}")

print(f"{Cores.purple}+{'='*largura}+{Cores.reset}")